[Go] Paging using channels

mcardosos opened the pull request:
PTAL
@jhendrixMSFT @marstr

I don't like what I wrote for the Lazy operation documentation :/ There must be a better way to write it.

marstr commented on src/generator/AutoRest.Go/Templates/MethodTemplate.cshtml:
```diff
@@ -193,3 +196,70 @@ func (client @(Model.Owner)) @(Model.ResponderMethodName)(resp *http.Response) (
+    @:respChan := make(chan @((Model.ListElement.ModelType as SequenceTypeGo).GetElement), 1)
+    @:errChan := make(chan error, 1)
+
+    @:go func() {
```
I think in general, we can probably reduce the amount of complexity in this method. I'll chat about this with you offline.

mcardosos commented on src/generator/AutoRest.Go/Model/MethodGo.cs:
```diff
@@ -102,6 +103,19 @@ public string MethodParametersSignature
+        signature.Append(MethodParametersSignature);
+        signature.Append(MethodParametersSignature.Length > 0
+          ? ", stop <-chan struct{})"
+          : "stop <-chan struct{})");
```
I don't want it to collide with the cancel channel used in long running operations.

mcardosos commented on src/generator/AutoRest.Go/Model/MethodGo.cs:
```diff
@@ -112,8 +126,37 @@ public string MethodReturnSignature
+
     public string NextMethodName => $"{Name}NextResults";
 
+    public string LazyMethodName => $"{Name}Lazy";
```
There are already some `ListAll` funcs  in the swaggers (they are for list by subscription instead than by resource group). To not erpeat funcs names, I chose `ListComplete`, but we can discuss which is a better name.

mcardosos commented on src/generator/AutoRest.Go/Model/MethodGo.cs:
```diff
@@ -365,15 +408,30 @@ public Response ReturnValue()
+          return methods.Any(m => m.Name.Value.EqualsIgnoreCase(next));
+      }
+      return false;
+      // return string.IsNullOrEmpty(NextMethodName);
```
fixed

mcardosos commented:
https://github.com/Azure/autorest/pull/2186
I have to remove stuff! :D

jhendrixMSFT commented:
Do you have some sample codegen that we can review?

mcardosos commented:
https://github.com/mcardosos/azure-sdk-for-go/blob/paging/arm/graphrbac/applications.go#L303

@jhendrixMSFT There it is.
Right now I can gen like this. I want to gen `Next` funcs that take the list result as receiver too (WIP).

mcardosos commented:
We happily talked about not including the `model.Next()` thing here, I have updated the PR to only include the `GetTheUniverse()` func. I also added it to all paging acceptance tests.
PTAL @marstr @jhendrixMSFT

jhendrixMSFT commented:
Do you have an updated sample of the generated code?

mcardosos commented:
Generated code for the acceptance tests looks like this...
``` go

// GetMultiplePagesComplete gets all elements from the list without paging.
func (client PagingClient) GetMultiplePagesComplete(clientRequestID string, maxresults *int32, timeout *int32, cancel <-chan struct{}) (<-chan Product, <-chan error) {
	resultChan := make(chan Product)
	errChan := make(chan error, 1)
	go func() {
		var err error
		defer func() {
			errChan <- err
			close(resultChan)
			close(errChan)
		}()
		list, err := client.GetMultiplePages(clientRequestID, maxresults, timeout)
		if err != nil {
			return
		}
		if list.Values != nil {
			for _, item := range *list.Values {
				select {
				case <-cancel:
					return
				default:
					resultChan <- item
				}
			}
		}
		for list.NextLink != nil {
			list, err = client.GetMultiplePagesNextResults(list)
			if err != nil {
				return
			}
			if list.Values != nil {
				for _, item := range *list.Values {
					select {
					case <-cancel:
						return
					default:
						resultChan <- item
					}
				}
			}
		}
	}()
	return resultChan, errChan
}
```

jhendrixMSFT commented on src/generator/AutoRest.Go/Templates/MethodTemplate.cshtml:
```diff
@@ -229,3 +211,59 @@ func (client @(Model.Owner)) @(Model.ResponderMethodName)(resp *http.Response) (
+    @:go func() {
+        @:var err error
+        @:defer func() {
+            @:errChan <- err
```
I think you want to change this to write to errChan only if err is not nil; this will allow the select on channels pattern to work properly (I made the same change in the generator for LROs).

mcardosos replied:
Fixed

marstr commented on src/generator/AutoRest.Go/Templates/MethodTemplate.cshtml:
```diff
@@ -229,3 +211,59 @@ func (client @(Model.Owner)) @(Model.ResponderMethodName)(resp *http.Response) (
+                    @:return
+                @:default:
+                    @:resultChan <- item
+                @:}
```
As written, the above will never really consider cancellation. Instead of using "default" the code we generate should look like this:

``` Go
for _, item := range *list.Items {
    select {
    case <- cancel:
        return
    case resultChan <- item:
        // Intentionally Left Blank
    }
}
```

mcardosos replied:
Fixed

mcardosos commented:
Generated code now looks like this...
```go
// GetMultiplePagesComplete gets all elements from the list without paging.
func (client PagingClient) GetMultiplePagesComplete(clientRequestID string, maxresults *int32, timeout *int32, cancel <-chan struct{}) (<-chan Product, <-chan error) {
	resultChan := make(chan Product)
	errChan := make(chan error, 1)
	go func() {
		defer func() {
			close(resultChan)
			close(errChan)
		}()
		list, err := client.GetMultiplePages(clientRequestID, maxresults, timeout)
		if err != nil {
			errChan <- err
			return
		}
		if list.Values != nil {
			for _, item := range *list.Values {
				select {
				case <-cancel:
					return
				default:
					resultChan <- item
				}
			}
		}
		for list.NextLink != nil {
			list, err = client.GetMultiplePagesNextResults(list)
			if err != nil {
				errChan <- err
				return
			}
			if list.Values != nil {
				for _, item := range *list.Values {
					select {
					case <-cancel:
						return
					case resultChan <- item:
						// Intentionally left blank
					}
				}
			}
		}
	}()
	return resultChan, errChan
}
```

marstr commented on src/generator/AutoRest.Go/Model/MethodGo.cs:
```diff
@@ -79,6 +79,78 @@ internal void Transform(CodeModelGo cmg)
+            get
+            {
+                StringBuilder sb = new StringBuilder();
+                foreach (var parameter in LocalParameters)
```
I don't see a new line being added into the documentation anywhere. I think it makes sense to have each parameter have its own line.

marstr commented on src/generator/AutoRest.Go/Model/MethodGo.cs:
```diff
@@ -79,6 +79,78 @@ internal void Transform(CodeModelGo cmg)
+                    }
+                }
+                return null;
+            }
```
I think we can simplify this method somewhat. What about this?

``` C#
get
{
    var body = ReturnType.Body as CompositeTypeGo
    return PropertyGo(body.Properties.Where(p => p is SequenceTypeGo).FirstOrDefault())
}
````

marstr commented on src/generator/AutoRest.Go/Model/MethodGo.cs:
```diff
@@ -402,15 +480,56 @@ public Response ReturnValue()
+                {
+                    return NextMethod.Name.Value;
+                }
+                return null;
```
The above can be simplified to:
``` C#
get
{
    return NextMethod?.Name.Value
}
```

mcardosos commented:
@azuresdkci retest this please
